vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java

```diff
@@ -400,6 +406,354 @@ public ElementQuery<T> withoutTheme(String theme) {
+        return withPlaceholder(placeholder, String::equals);
+    }
+
+    /**
+     * Requires the element's placeholder to partially match the given text value.
+     * <p>
+     * For exactly matching the placeholder, see
+     * {@link #withPlaceholder(String)}.
+     *
+     * @param text
+     *          the text to match
+     * @return this element query instance for chaining
+     *
+     * @see #withPlaceholder(String)
+     * @see #withPlaceholder(String, BiPredicate)
+     */
+    public ElementQuery<T> withPlaceholderContaining(String text) {
+        return withPlaceholder(text, String::contains);
+    }
+
+    /**
+     * Requires the element's caption (i.e., label, placeholder, or text label)
+     * to satisfy the given comparison with the supplied text.
+     * <p>
+     * If the element has a label, the label value is used.
+     * If the element can have a label, but its value is empty and it has a placeholder,
+     * the placeholder value is used.
+     * If the element does not support labels nor placeholders
+     * but does support labels via its text (such as a button),
+     * the text is used.
+     * <p>
+     * For matching a caption exactly, see {@link #withCaption(String)},
+     * and for matching a caption partially, see {@link #withCaptionContaining(String)}.
+     * <p>
+     * This method can be used for other caption matching needs,
+     * such as performing a case-insensitive match:
+     * {@code withCaption("name", String::equalsIgnoreCase)}
+     * <br> a caption-ending match:
+     * {@code withCaption(" Name", String::endsWith)}
+     * <br> or a regular expression match:
+     * {@code withCaption("(First|Last) Name", String::matches)}
+     *
+     * @param text
+     *          the text to compare with the caption
+     * @param comparison
+     *          the comparison to use when comparing the text with the caption
+     * @return this element query instance for chaining
+     *
+     * @see #withCaption(String)
+     * @see #withCaptionContaining(String)
+     */
+    public ElementQuery<T> withCaption(String text, BiPredicate<String, String> comparison) {
+        Objects.requireNonNull(comparison, "comparison must not be null");
+        return withCondition(element -> {
+            if (element instanceof HasLabel hasLabel) {
+                var label = hasLabel.getLabel();
+                if (!label.isEmpty()) {
+                    return comparison.test(label, text);
+                }
+            }
+            if (element instanceof HasPlaceholder hasPlaceholder) {
+                var placeholder = hasPlaceholder.getPlaceholder();
+                if (!placeholder.isEmpty()) {
+                    return comparison.test(placeholder, text);
+                }
+            }
+            if (element instanceof HasLabelAsText labelAsText) {
+                var label = labelAsText.getText();
+                if (!label.isEmpty()) {
```

Comment:
Looking at the Javadoc of `WebElement.getText()` there is no guarantee that the returned value is not null, so maybe better add a null check